### ISSUE TITLE
Is there a function to profile the onnx file?

### ISSUE THREAD

**Issue by ohjunee** [ultralytics/ultralytics#4509]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

The yolo model is being profiled by referring to the torch_utils.py code.
The current pytorch-type file is implemented as follows.

```python
import torchvision.models as models
from ultralytics import YOLO
from ultralytics.utils.torch_utils import * 

resnet18 = models.resnet18().to('cuda:0')
mobilenet_v2 = models.mobilenet_v2().to('cuda:0')
yolov8s = YOLO('yolov8s.pt').model.to('cuda:0')
yolov8m = YOLO('yolov8m.pt').model.to('cuda:0')
models = [resnet18, mobilenet_v2, yolov8s, yolov8m]
model_names = ['Resnet-18', 'MobileNet-v2', 'Yolov8s', 'Yolov8m']

for (model, name) in zip(models, model_names):
    model.eval()
    # model_info(model, detailed=False)
    if name in ['Resnet-18', 'MobileNet-v2']:
        print(f'{name}: {get_flops_with_torch_profiler(model, imgsz=224)}GFLOPS')
    else:
        print(f'{name}: {get_flops_with_torch_profiler(model)}GFLOPS')
``` 

Is there a function to profile the onnx file? 
From what I looked for, torch_utils.The model_info_for_loggers() of py seems relevant.

```python
def model_info_for_loggers(trainer):
    """
    Return model info dict with useful model information.

    Example for YOLOv8n:
        {'model/parameters': 3151904,
         'model/GFLOPs': 8.746,
         'model/speed_ONNX(ms)': 41.244,
         'model/speed_TensorRT(ms)': 3.211,
         'model/speed_PyTorch(ms)': 18.755}
    """
    if trainer.args.profile:  # profile ONNX and TensorRT times
        from ultralytics.utils.benchmarks import ProfileModels
        results = ProfileModels([trainer.last], device=trainer.device).profile()[0]
        results.pop('model/name')
    else:  # only return PyTorch times from most recent validation
        results = {
            'model/parameters': get_num_params(trainer.model),
            'model/GFLOPs': round(get_flops(trainer.model), 3)}
    results['model/speed_PyTorch(ms)'] = round(trainer.validator.speed['inference'], 3)
    return results
``` 

Thank you.

**Comment by glenn-jocher** [comment]
@ohjunee hi there,

Yes, indeed we do support profiling for models saved in the PyTorch (*.pt) format. While your code currently profiles models in the ONNX format, you can directly use the PyTorch models for the profiling as well. The current implementation should work seamlessly with *.pt files. Just replace your model list with the names of your *.pt model files.

In this regard, remember to ensure that the models are correctly loaded on the device specified (in your case, the GPU with id 0). You might also want to consider whether using half precision (FP16) is appropriate for your use case, because this can significantly speed up the model inference, depending on the GPU you are using.

Let's keep improving YOLOv8 together. Please let us know if you have additional questions or if you run into any issues.

**Comment by ohjunee** [comment]
For yaml, pt, and yml types of inputs, we check the model information in the pytorch file, but it seems that only onnx and trt are provided for the inference speed. Is it possible to check the information about the inference speed of the pt file?

```python
            if file.suffix in ('.pt', '.yaml', '.yml'):
                model = YOLO(str(file))
                model.fuse()  # to report correct params and GFLOPs in model.info()
                model_info = model.info()
                if self.trt and self.device.type != 'cpu' and not engine_file.is_file():
                    engine_file = model.export(format='engine',
                                               half=True,
                                               imgsz=self.imgsz,
                                               device=self.device,
                                               verbose=False)
                onnx_file = model.export(format='onnx',
                                         half=half,
                                         imgsz=self.imgsz,
                                         simplify=simplify,
                                         device=self.device,
                                         verbose=False)
            elif file.suffix == '.onnx':
                model_info = self.get_onnx_model_info(file)
                onnx_file = file
            else:
                continue

            t_engine = self.profile_tensorrt_model(str(engine_file))
            t_onnx = self.profile_onnx_model(str(onnx_file))
            table_rows.append(self.generate_table_row(file.stem, t_onnx, t_engine, model_info))
            output.append(self.generate_results_dict(file.stem, t_onnx, t_engine, model_info))
```

**Comment by ohjunee** [comment]
> 안녕
> 
> 예, 실제로 PyTorch(*.pt) 형식으로 저장된 모델에 대한 프로파일링을 지원합니다. 코드는 현재 ONNX 형식으로 모델을 프로파일링하지만 프로파일링에 PyTorch 모델을 직접 사용할 수도 있습니다. 현재 구현은 *.pt 파일과 원활하게 작동해야 합니다. 모델 목록을 *.pt 모델 파일의 이름으로 바꾸기만 하면 됩니다.
> 
> 이와 관련하여 모델이 지정된 장치 (귀하의 경우 ID가 0 인 GPU)에 올바르게로드되었는지 확인하십시오. 또한 사용 중인 GPU에 따라 모델 추론 속도를 크게 높일 수 있으므로 반정밀도(FP16)를 사용하는 것이 사용 사례에 적합한지 여부를 고려할 수도 있습니다.
> 
> YOLOv8을 함께 개선해 나갑시다. 추가 질문이 있거나 문제가 발생하면 알려주십시오.

For yaml, pt, and yml types of inputs, we check the model information in the pytorch file, but it seems that only onnx and trt are provided for the inference speed. Is it possible to check the information about the inference speed of the pt file?

```python
            if file.suffix in ('.pt', '.yaml', '.yml'):
                model = YOLO(str(file))
                model.fuse()  # to report correct params and GFLOPs in model.info()
                model_info = model.info()
                if self.trt and self.device.type != 'cpu' and not engine_file.is_file():
                    engine_file = model.export(format='engine',
                                               half=True,
                                               imgsz=self.imgsz,
                                               device=self.device,
                                               verbose=False)
                onnx_file = model.export(format='onnx',
                                         half=half,
                                         imgsz=self.imgsz,
                                         simplify=simplify,
                                         device=self.device,
                                         verbose=False)
            elif file.suffix == '.onnx':
                model_info = self.get_onnx_model_info(file)
                onnx_file = file
            else:
                continue

            t_engine = self.profile_tensorrt_model(str(engine_file))
            t_onnx = self.profile_onnx_model(str(onnx_file))
            table_rows.append(self.generate_table_row(file.stem, t_onnx, t_engine, model_info))
            output.append(self.generate_results_dict(file.stem, t_onnx, t_engine, model_info))
```

**Comment by glenn-jocher** [comment]
@ohjunee yes, you got it correct. We can indeed evaluate the performance of PyTorch (*.pt) models using the `benchmark()` function available. 

This function works by running a benchmark test across different formats (in your case, PyTorch format *.pt) to evaluate speed and accuracy. 

It allows you to evaluate a YOLOv8 model on a specific dataset by benchmarking/frame-working it at a particular image size. 

You also have the option to specify the precision to be used (half-precision or int8-precision), and the device to run your benchmark test on, either on a 'cpu' or 'cuda' (GPU).

Moreover, the function returns a pandas DataFrame with benchmark results for each format, including file size, metric, and inference time, which could be very helpful in the analysis.

I hope this helps! Please let me know if you have any more questions or if anything is not clear. Let's keep improving YOLOv8 together!
